### ISSUE TITLE
cnaProduce and gapReport fail on combineMetaCycKEGGModels 

### ISSUE THREAD

**Issue by phemmy2k2** [SysBioChalmers/RAVEN#132]
I combined metacyc and kegg models, then I discovered I can not apply the canProduce and gapReport function on the combined model. The following is the output error.

#### Reproducing this issue:
```matlab
>> CombinedDraftModel=combineMetaCycKEGGModels(model_MetaCyc, balancedModel);

>> I=canProduce(CombinedDraftModel);
Matrix index is out of range for deletion.

Error in removeReactions (line 68)
            reducedModel.grRules(indexesToDelete,:)=[];

Error in simplifyModel (line 148)
            reducedModel=removeReactions(reducedModel,rxnsToDelete);

Error in haveFlux (line 45)
smallModel=simplifyModel(model,false,false,true,true);

Error in canProduce (line 25)
produced=haveFlux(model,10^-5,rxns);
 
>> [noFluxRxns, noFluxRxnsRelaxed, subGraphs, notProducedMets, minToConnect,...
    neededForProductionMat]=gapReport(CombinedDraftModel);
Gap analysis for COMBINED - Combined model from MetaCyc and KEGG draft models

Matrix index is out of range for deletion.

Error in removeReactions (line 68)
            reducedModel.grRules(indexesToDelete,:)=[];

Error in simplifyModel (line 148)
            reducedModel=removeReactions(reducedModel,rxnsToDelete);

Error in haveFlux (line 45)
smallModel=simplifyModel(model,false,false,true,true);

Error in gapReport (line 66)
I=haveFlux(model);
```
#### System information
* Please report:
1. RAVEN version 2.0 
2. Windows Operating system 
**I hereby confirm that I have:**
- [X] Followed the [guidelines](https://github.com/SysBioChalmers/RAVEN/wiki/Installation) to install RAVEN.
- [X] Checked that a similar issue does not [already exist](https://github.com/SysBioChalmers/RAVEN/issues?utf8=%E2%9C%93&q=is%3Aissue)
- [X] If suitable, needed, asked first in the [Gitter chat room](https://gitter.im/SysBioChalmers/RAVEN) about the issue

**Comment by edkerk** [comment]
Something seems wrong with `CombinedDraftModel.grRules`, having less entries than the number of reactions? I cannot replicate this with my own code, can you send the CombinedDraftModel? Also, how did you construct `model_MetaCyc` and `balancedModel` (do they have the same length of grRules as the number of rxns?).

**Comment by phemmy2k2** [comment]
@edkerk, How do you want to receive the combinedDraftModel? I can also send the complete code to understand how I construct model_MetaCyc and balancedModel.

**Comment by phemmy2k2** [comment]
Here is a detailed code. 

% Get model from MetaCyc
model_MetaCyc = getMetaCycModelForOrganism('ano2cyc','AgamP3_protein.faa',true,false,false)

% Draft GEM reconstructed from homology with HMMs traind from KEGG
model_KEGG=getKEGGModelForOrganism('aga','AgamP3_protein.faa','euk100_kegg82','output',false,false,false,10^-15);

% Remove bad reactions from model_KEGG but IGNORE H+ to output a newModel
[newModel, removedRxns]=removeBadRxns(model_KEGG,1,{'H+'},true);

% Add some uptakes and see what it can produce.
[I, J]=ismember({'Sucrose';'H2O';'Heme';'Oxygen';'Glutathione';'NAD'},newModel.metNames);
[newModel, addedRxns]=addExchangeRxns(newModel,'in',J);

% Check which metabolites can be produced given these uptakes
%This throws error for combined KEGG-MetaCyc model
I=canProduce(newModel);
sum(I)/numel(newModel.mets)

% We can try to fill gaps using the full KEGG model 
keggModel=getModelFromKEGG([],false,false,false);

% Let's only use the balanced ones for the gap filling
balanceStructure=getElementalBalance(keggModel);
keggModel=removeReactions(keggModel,balanceStructure.balanceStatus~=1,true,true);

params.relGap=0.6; 
params.printReport=true;
[newConnected, cannotConnect, addedRxns, balancedModel, exitFlag]=fillGaps(newModel,keggModel,true,false,false,[],params);

% I Combined the KEGG and MetaCyc draft GEMs after fillGaps because the fillGaps throws error %for merged model
CombinedDraftModel=combineMetaCycKEGGModels(model_MetaCyc, balancedModel);

[noFluxRxns, noFluxRxnsRelaxed, subGraphs, notProducedMets, minToConnect,...
 neededForProductionMat]=gapReport(CombinedDraftModel);

**Comment by edkerk** [comment]
@phemmy2k2 thanks for the script, I'll try to replicate and get back to you, will take some days.

**Comment by phemmy2k2** [comment]
@edkerk I am checking on you to know if you have replicated the problem.

**Comment by edkerk** [comment]
@phemmy2k2 Summer holidays delayed my response, but I'm currently looking into this!

**Comment by edkerk** [comment]
As a quick-fix, run `CombinedDraftModel=rmfield(CombinedDraftModel,'equations');` right after you ran `combineMetaCycKEGGModels`. The 'equations' field is problematic, and you'll probably won't need this field, you can just remove it.
@Hao-Chalmers: is the 'equations' field necessary?

**Comment by Hao-Chalmers** [comment]
@edkerk In RAVEN structure, 'equations' field is optional. I think it is fine to have it, and user is free to remove it. The question is whether this field caused this bug?

**Comment by edkerk** [comment]
@Hao-Chalmers It caused this bug, removing it solves the problem. The issue is that `addExchangeRxns` doesn't deal with 'equations' fields (it should add empty cells), but then the question might become how many functions we should adapt for compatibility with the 'equations' field if that field is not used anywhere else.

**Comment by Hao-Chalmers** [comment]
The right way is to modify `addExchangeRxns` by adding additional checks to 'equation' field. And we should start a branch to systematically resolve functions with this similar issue.

**Comment by phemmy2k2** [comment]
Thank you @edkerk  On removing the equations field the canProduce() completed successfully, however there are still some errors using the fillgaps() on the combined model.

>> [newConnected, cannotConnect, addedRxns, balancedModel, exitFlag]=fillGaps(CombinedDraftModel,keggModel,true,false,false,[],params);
Matrix index is out of range for deletion.

Error in removeMets (line 101)
        reducedModel.metCharges(indexesToDelete)=[];

Error in simplifyModel (line 154)
            reducedModel=removeMets(reducedModel,notInUse);

Error in fillGaps (line 181)
    allModels=simplifyModel(allModels,false,false,false,true,false,false,false,[],true);

**Comment by edkerk** [comment]
Thanks @phemmy2k2 for reporting this bug in mergeModels, this is now fixed in commit ecfa74b9b3ead9ff17aade4edf727487bceb0a69. I will close this issue, feel free to open a new one if you run into any other bugs or issues!
